### PR TITLE
Wifi-888 : Channel switch events due to high interference and DFS

### PR DESCRIPTION
**Pull request by AmmadRehmatCUS**
Wifi-888 : Channel switch events due to high interference and DFS

Signed-off-by: Ammad Rehmat <ammad.rehmat@connectus.ai>

### CODE REVIEW

**Review by yashvardhan-netex — COMMENTED** [review]
It seems there are multiple issues with the patch. Would be great if you can share what all tests were done. Also if you can post the ubus reply for get_chan_switch_events after switching the channel.

File: feeds/wifi-trunk/hostapd/patches/901-hapd-ubus-channel-switch.patch [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
++
++	struct hostapd_chan_event_list *entry;
++	blob_buf_init(&b_ev, 0);
++	a = blobmsg_open_table(&b_ev, "chan_switch_event");
++	list_for_each_entry(entry, &chan_events->list, list)
++	{
++		t = blobmsg_open_table(&b_ev, "event");
++		blobmsg_add_u32(&b_ev, "radio_name", entry->records.band);
++		blobmsg_add_u8(&b_ev, "reason", entry->records.reason);
++		blobmsg_add_u64(&b_ev, "timestamp", entry->records.ts.tv_sec * (uint64_t)1000000);
```
Comment (yashvardhan-netex) [inline feeds/wifi-trunk/hostapd/patches/901-hapd-ubus-channel-switch.patch:57]: This way we will not be able to achieve the targeted milliseconds or microseconds precision. tv_nsec (which holds the rest of the elapsed time in nanoseconds) needs to be used in addition to tv_sec ofcourse with the desired ms/us conversion.

File: feeds/wlan-ap/opensync/patches/26-channel-switch-events.patch [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+@@ -184,6 +187,22 @@ typedef struct {
+ 	ds_dlist_node_t node;
+ } dpp_event_record_t;
+ 
++/*proto: ChannelSwitchEvent*/
++
++typedef struct {
++       radio_type_t    band;
++       cs_reason_t     reason;
++       uint8_t         freq;
```
Comment (yashvardhan-netex) [inline feeds/wlan-ap/opensync/patches/26-channel-switch-events.patch:67]: I think we are trying to send the changed channel to cloud. This should be channel and not frequency

File: feeds/wlan-ap/opensync/src/src/sm/src/sm_events_report.c [hunk 3]
```diff
@@ -134,12 +134,13 @@ static void sm_events_report(EV_P_ ev_timer *w, int revents)
...
 	}
 
-	if (!ds_dlist_is_empty(&report_ctx->client_event_list)) {
+	if (!ds_dlist_is_empty(&report_ctx->client_event_list) || !ds_dlist_is_empty(&report_ctx->channel_switch_list)) {
 		LOG(INFO, "Sending events report...");
 		dpp_put_events(report_ctx);
 	}
 
 	sm_events_report_clear(&report_ctx->client_event_list);
+	sm_events_report_clear(&report_ctx->channel_switch_list);
```
Comment (yashvardhan-netex) [inline feeds/wlan-ap/opensync/src/src/sm/src/sm_events_report.c:143]: This will not work. "sm_events_report_clear" only deletes client event records. You would have to extend it for channel switch events

File: feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c [hunk 4]
```diff
@@ -18,6 +18,26 @@ typedef struct {
...
 } delete_entry_t;
 
+typedef struct {
+       char        *radio_name;
+       uint8_t     reason;
+       uint8_t     freq;
+       uint32_t    timestamp;
+
+       ds_dlist_node_t node;	
+}delete_entry_chan_event_t;
```
Comment (yashvardhan-netex) [inline feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c:28]: This is not used anywhere and do we need it?

File: feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c [hunk 5]
```diff
@@ -541,6 +561,94 @@ static void ubus_collector_bss_cb(struct ubus_request *req, int type,
...
+			    "ubus_collector_ failed while parsing chan event policy");
+			continue;
+		}
+
+		channel_event_record = malloc(sizeof(dpp_event_channel_switch_t));
+
+		channel_event_record->channel_event.band = frequency_to_band(blobmsg_get_u32(tb_chan_event_table[CHANNEL_SWITCH_RADIO_NAME]));
+
+		channel_event_record->channel_event.reason = blobmsg_get_u8(tb_chan_event_table[CHANNEL_SWITCH_REASON]);
+		channel_event_record->channel_event.freq = blobmsg_get_u8(tb_chan_event_table[CHANNEL_SWITCH_FREQ]);
```
Comment (yashvardhan-netex) [inline feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c:607]: Missing frequency to channel conversion. Further, u8 is not sufficient to hold frequency.

File: feeds/wlan-ap/opensync/patches/26-channel-switch-events.patch [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
++    /* Multiple Channel Switch events */
++    repeated ChannelSwitchEvent         channel_switch                      = 2;
++
+     /* Multiple DHCP Transactions */
+     repeated DhcpTransaction            dhcp_transaction                    = 3;
+ }
+Index: opensync-2.0.5.0/src/lib/datapipeline/inc/dpp_events.h
+===================================================================
+--- opensync-2.0.5.0.orig/src/lib/datapipeline/inc/dpp_events.h
++++ opensync-2.0.5.0/src/lib/datapipeline/inc/dpp_events.h
```
Comment (yashvardhan-netex) [inline feeds/wlan-ap/opensync/patches/26-channel-switch-events.patch:47]: This should not be part of the patch. dpp_events.h is a git revision file

**Review by yashvardhan-netex — COMMENTED** [review]
May be it would be good to also report the previous channel from which the switch happened

File: feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c [hunk 7]
```diff
@@ -541,6 +561,94 @@ static void ubus_collector_bss_cb(struct ubus_request *req, int type,
...
+	ubus_invoke_async(ubus, ubus_object_id, hostapd_method, NULL, req);
+
+	req->data_cb = (ubus_data_handler_t)ubus_collector_chan_switch_events_cb;
+	req->complete_cb = (ubus_complete_handler_t)ubus_collector_complete_channel_switch_cb;
+
+	ubus_complete_request(ubus, req, 0);
+
+	evsched_task_reschedule();
+}
+
```
Comment (yashvardhan-netex) [inline feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c:651]: This will not work. "get_chan_switch_events" is defined per bss. ubus object path should be hostapd.wlan0, hostapd.wlan1 and so on and not hostapd. This needs to be called for each bss.

**Review by yashvardhan-netex — APPROVED** [review]
Looks good to me.

**Comment by RickSommerville** [comment]
Merged, thanks
